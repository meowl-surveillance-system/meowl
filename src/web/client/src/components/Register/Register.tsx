import React, {
  ChangeEvent,
  Component,
  ComponentState,
  FormEvent,
} from "react";
import { Redirect, Link as RouterLink } from "react-router-dom";
import Button from "@material-ui/core/Button";
import TextField from "@material-ui/core/TextField";
import Grid from "@material-ui/core/Grid";
import Typography from "@material-ui/core/Typography";
import Container from "@material-ui/core/Container";

interface Props {
  isLoggedIn: boolean;
}
interface State {
  email: string;
  username: string;
  password: string;
  registeredBadMessage: string;
  registeredGoodMessage: string;
}

/**
 * A register form for users to register an account
 */
export default class Register extends Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      email: "",
      username: "",
      password: "",
      registeredBadMessage: "",
      registeredGoodMessage: "",
    };
  }

  /**
   * Sets the appropriate state when a keyboard event is triggered
   * @params event - An event that records changes happening to an input field
   */
  onChange = (event: ChangeEvent<HTMLInputElement>): void => {
    event.preventDefault();
    this.setState({
      [event.target.name]: event.target.value,
    } as ComponentState);
  };

  /**
   * Submits the user credentials to the auth server
   * @params event - An event that records submission of a form element
   *
   * Changes the root level isLoggedIn state to True and redirects to the
   * "/streams" route if register submission is successful.
   * Otherwise, alert the user that register is not successful.
   */
  onSubmit = async (event: FormEvent<HTMLFormElement>) => {
    try {
      event.preventDefault();
      const res = await fetch("/auth/register", {
        method: "POST",
        headers: {
          Accept: "application/json",
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          email: this.state.email,
          username: this.state.username,
          password: this.state.password,
        }),
      });
      const msg = await res.text();
      if(msg === 'successfully added to pending accounts') {
        this.setState({ registeredGoodMessage: msg, registeredBadMessage: "" });
      }
      else {
        this.setState({ registeredBadMessage: msg, registeredGoodMessage: "" });
      }
    } catch (error) {
      console.error(error);
    }
  };

  /**
   * Redirect user to "/" if user is already logged in
   * Otherwise, render the register form
   */
  render() {
    if (this.props.isLoggedIn) {
      return <Redirect to="/" />;
    }
    return (
      <Container component="main" maxWidth="xs">
        <Grid
          container
          spacing={2}
          direction="column"
          alignItems="center"
          justify="center"
          style={{ minHeight: "100vh" }}
        >
          {this.state.registeredBadMessage && <Typography style={{color: "red"}}>{this.state.registeredBadMessage}</Typography>}
          {this.state.registeredGoodMessage && <Typography style={{color: "green"}}>{this.state.registeredGoodMessage}</Typography>}
          <Typography component="h1" variant="h5">
            Register
          </Typography>
          <form noValidate onSubmit={this.onSubmit}>
            <TextField
              variant="outlined"
              margin="normal"
              required
              fullWidth
              label="E-mail"
              name="email"
              onChange={this.onChange}
            />
            <TextField
              variant="outlined"
              margin="normal"
              required
              fullWidth
              label="Username"
              name="username"
              onChange={this.onChange}
            />
            <TextField
              variant="outlined"
              margin="normal"
              required
              fullWidth
              name="password"
              label="Password"
              type="password"
              onChange={this.onChange}
            />
            <Button fullWidth type="submit" variant="contained" color="primary">
              Submit
            </Button>
            <Grid container>
              <Grid item>
                <RouterLink to="/login">
                  {"Already have an account? Sign In"}
                </RouterLink>
              </Grid>
            </Grid>
          </form>
        </Grid>
      </Container>
    );
  }
}
