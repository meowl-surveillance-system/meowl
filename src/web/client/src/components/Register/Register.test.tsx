import React, { FormEvent } from "react";
import { shallow } from "enzyme";
import { render } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import Register from "./Register";

describe("Register component", () => {
  let registerWrapper: any;
  let onAuthChangeMock: jest.Mock;
  let historyMock: object;
  let eventMock: any;
  let preventDefaultMock: jest.Mock;
  beforeEach(() => {
    // Mock Fetch
    const mockSuccessResponse = "successfully registered";
    const mockTextPromise: any = Promise.resolve(mockSuccessResponse);
    const mockFetchPromise: any = Promise.resolve({
      text: () => mockTextPromise,
    });
    jest.spyOn(window, "fetch").mockImplementation(() => mockFetchPromise);

    preventDefaultMock = jest.fn();
    eventMock = {
      preventDefault: preventDefaultMock,
      target: {
        name: "email",
        value: "test@test.com",
      },
    };

    historyMock = {
      push: jest.fn(),
    };

    onAuthChangeMock = jest.fn();

    registerWrapper = shallow(
      <Register
        isLoggedIn={false}
        onAuthChange={onAuthChangeMock}
        history={historyMock}
      />,
    );
  });

  it("renders Register component", () => {
    const { container } = render(
      <Router>
        <Register
          isLoggedIn={false}
          onAuthChange={onAuthChangeMock}
          history={historyMock}
        />
      </Router>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should call preventDefault on onSubmit", async () => {
    await registerWrapper.instance().onSubmit(eventMock);
    expect(preventDefaultMock).toHaveBeenCalled();
  });

  it("should call onAuthChange on onSubmit", async () => {
    await registerWrapper.instance().onSubmit(eventMock);
    expect(onAuthChangeMock).toHaveBeenCalled();
  });

  it("should set email field when onChange is called", () => {
    registerWrapper.instance().onChange(eventMock);
    expect(registerWrapper.instance().state.email).toEqual(
      eventMock.target.value,
    );
  });
});
