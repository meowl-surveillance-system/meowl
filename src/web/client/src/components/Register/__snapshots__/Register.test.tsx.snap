// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register component renders Register component 1`] = `
<main
  class="MuiContainer-root MuiContainer-maxWidthXs"
>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
    style="min-height: 100vh;"
  >
    <h1
      class="MuiTypography-root MuiTypography-h5"
    >
      Register
    </h1>
    <form
      novalidate=""
    >
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
          data-shrink="false"
        >
          E-mail
          <span
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input"
            name="email"
            required=""
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-200 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-202"
            >
              <span>
                E-mail
                 *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
          data-shrink="false"
        >
          Username
          <span
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input"
            name="username"
            required=""
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-200 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-202"
            >
              <span>
                Username
                 *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
          data-shrink="false"
        >
          Password
          <span
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input"
            name="password"
            required=""
            type="password"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-200 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-202"
            >
              <span>
                Password
                 *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-fullWidth"
        tabindex="0"
        type="submit"
      >
        <span
          class="MuiButton-label"
        >
          Submit
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <div
        class="MuiGrid-root MuiGrid-container"
      >
        <div
          class="MuiGrid-root MuiGrid-item"
        >
          <a
            href="/login"
          >
            Already have an account? Sign In
          </a>
        </div>
      </div>
    </form>
  </div>
</main>
`;
